From b951988aad563d7966e6971aa0917751f1520e88 Mon Sep 17 00:00:00 2001
From: Nic Costa <nic.costa@arm.com>
Date: Wed, 11 Oct 2017 13:28:50 -0500
Subject: [PATCH 1/2] Support for external SPI Flash

* Add spif-driver.lib
* Use the SPIFBlockDevice driver instead of SDBlockDevice if
  configured in mbed_app.json

Signed-off-by: Nic Costa <nic.costa@arm.com>
---
 mbed_app.json   |  6 ++++++
 source/main.cpp | 19 +++++++++++++------
 spif-driver.lib |  1 +
 3 files changed, 20 insertions(+), 6 deletions(-)
 create mode 100644 spif-driver.lib

diff --git a/mbed_app.json b/mbed_app.json
index 80f87c1..d0d582e 100755
--- a/mbed_app.json
+++ b/mbed_app.json
@@ -19,6 +19,11 @@
         "serial_baudrate": {
             "help": "Terminal default baudrate",
             "value": 115200
+        },
+        "spif-spi-cs": "D10",
+        "use-spi-flash": {
+            "help": "Set to true to use SPI Flash or false to use SD card.",
+            "value": "true"
         }
     },
     "target_overrides": {
@@ -28,6 +33,7 @@
             "platform.stdio-baud-rate": 115200
         },
         "K64F": {
+            "app.use-spi-flash": false,
             "firmware_metadata_header_address": "0x20000",
             "update-client.application-details": "0x20000",
             "firmware_metadata_header_size": "0x400"
diff --git a/source/main.cpp b/source/main.cpp
index 2a37a7e..d150da2 100755
--- a/source/main.cpp
+++ b/source/main.cpp
@@ -24,7 +24,11 @@
 #include "update-client-pal-filesystem/arm_uc_pal_filesystem.h"
 #include "update-client-paal/arm_uc_paal_update.h"
 #include "update-client-common/arm_uc_types.h"
+#if MBED_CONF_APP_USE_SPI_FLASH
+#include <SPIFBlockDevice.h>
+#else
 #include "SDBlockDevice.h"
+#endif
 #include "pal.h"
 
 #include "mbed_bootloader_info.h"
@@ -50,13 +54,16 @@ const arm_uc_installer_details_t bootloader = {
 };
 
 /* initialise sd card and filesystem */
-#if defined(MBED_CONF_APP_SPI_MOSI) && defined(MBED_CONF_APP_SPI_MISO) && \
-    defined(MBED_CONF_APP_SPI_CLK)  && defined(MBED_CONF_APP_SPI_CS)
-SDBlockDevice sd(MBED_CONF_APP_SPI_MOSI, MBED_CONF_APP_SPI_MISO,
-                 MBED_CONF_APP_SPI_CLK,  MBED_CONF_APP_SPI_CS);
+#if MBED_CONF_APP_USE_SPI_FLASH
+SPIFBlockDevice sd(MBED_CONF_SD_SPI_MOSI,
+                   MBED_CONF_SD_SPI_MISO,
+                   MBED_CONF_SD_SPI_CLK,
+                   MBED_CONF_APP_SPIF_SPI_CS);
 #else
-SDBlockDevice sd(MBED_CONF_SD_SPI_MOSI, MBED_CONF_SD_SPI_MISO,
-                 MBED_CONF_SD_SPI_CLK,  MBED_CONF_SD_SPI_CS);
+SDBlockDevice sd(MBED_CONF_SD_SPI_MOSI,
+                 MBED_CONF_SD_SPI_MISO,
+                 MBED_CONF_SD_SPI_CLK,
+                 MBED_CONF_SD_SPI_CS);
 #endif
 FATFileSystem fs("sd", &sd);
 
diff --git a/spif-driver.lib b/spif-driver.lib
new file mode 100644
index 0000000..f365260
--- /dev/null
+++ b/spif-driver.lib
@@ -0,0 +1 @@
+https://github.com/ARMmbed/spif-driver/#c46b2a762d03874eb712250c7849759f9b13a093
-- 
2.14.2

